// graphql-tools-fork/dist/stitching/delegateToSchema.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var Interfaces_1 = require("../Interfaces");
var transforms_1 = require("../transforms/transforms");
var AddArgumentsAsVariables_1 = __importDefault(require("../transforms/AddArgumentsAsVariables"));
var FilterToSchema_1 = __importDefault(require("../transforms/FilterToSchema"));
var AddTypenameToAbstract_1 = __importDefault(require("../transforms/AddTypenameToAbstract"));
var CheckResultAndHandleErrors_1 = __importDefault(require("../transforms/CheckResultAndHandleErrors"));
var mapAsyncIterator_1 = __importDefault(require("./mapAsyncIterator"));
var ExpandAbstractTypes_1 = __importDefault(require("../transforms/ExpandAbstractTypes"));
var AddReplacementFragments_1 = __importDefault(require("../transforms/AddReplacementFragments"));
var apollo_link_1 = require("apollo-link");
var linkToFetcher_1 = __importDefault(require("./linkToFetcher"));
var observableToAsyncIterable_1 = require("./observableToAsyncIterable");
var transforms_2 = require("../transforms");
function delegateToSchema(options) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    if (options instanceof graphql_1.GraphQLSchema) {
        throw new Error('Passing positional arguments to delegateToSchema is a deprecated. ' +
            'Please pass named parameters instead.');
    }
    return delegateToSchemaImplementation(options);
}
exports.default = delegateToSchema;
function delegateToSchemaImplementation(_a) {
    var subschema = _a.schema, rootValue = _a.rootValue, info = _a.info, _b = _a.operation, operation = _b === void 0 ? info.operation.operation : _b, fieldName = _a.fieldName, args = _a.args, context = _a.context, _c = _a.transforms, transforms = _c === void 0 ? [] : _c, skipValidation = _a.skipValidation;
    return __awaiter(this, void 0, void 0, function () {
        var targetSchema, subSchemaConfig, rawDocument, rawRequest, processedRequest, errors, executor, _d, subscriber, originalAsyncIterator;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    if (Interfaces_1.isSubschemaConfig(subschema)) {
                        subSchemaConfig = subschema;
                        targetSchema = subSchemaConfig.schema;
                        rootValue = rootValue || subSchemaConfig.rootValue || info.rootValue;
                        transforms = transforms.concat((subSchemaConfig.transforms || []).slice().reverse());
                    }
                    else {
                        targetSchema = subschema;
                        rootValue = rootValue || info.rootValue;
                    }
                    rawDocument = createDocument(fieldName, operation, info.fieldNodes, Object.keys(info.fragments).map(function (fragmentName) { return info.fragments[fragmentName]; }), info.operation.variableDefinitions, info.operation.name);
                    rawRequest = {
                        document: rawDocument,
                        variables: info.variableValues,
                    };
                    transforms = __spreadArrays([
                        new CheckResultAndHandleErrors_1.default(info, fieldName, subschema, context)
                    ], transforms, [
                        new ExpandAbstractTypes_1.default(info.schema, targetSchema),
                    ]);
                    if (info.mergeInfo) {
                        transforms.push(new AddReplacementFragments_1.default(targetSchema, info.mergeInfo.replacementFragments), new transforms_2.AddMergedTypeFragments(targetSchema, info.mergeInfo.mergedTypes));
                    }
                    if (args) {
                        transforms.push(new AddArgumentsAsVariables_1.default(targetSchema, args, info.schema));
                    }
                    else {
                        console.warn('"args" undefined. "args" argument may be required in a future version. Custom scalars or enums may not be properly serialized prior to delegation.');
                    }
                    transforms = transforms.concat([
                        new FilterToSchema_1.default(targetSchema),
                        new AddTypenameToAbstract_1.default(targetSchema),
                    ]);
                    processedRequest = transforms_1.applyRequestTransforms(rawRequest, transforms);
                    if (!skipValidation) {
                        errors = graphql_1.validate(targetSchema, processedRequest.document);
                        if (errors.length > 0) {
                            throw errors;
                        }
                    }
                    if (!(operation === 'query' || operation === 'mutation')) return [3 /*break*/, 2];
                    executor = createExecutor(targetSchema, rootValue, subSchemaConfig);
                    _d = transforms_1.applyResultTransforms;
                    return [4 /*yield*/, executor({
                            document: processedRequest.document,
                            context: context,
                            variables: processedRequest.variables
                        })];
                case 1: return [2 /*return*/, _d.apply(void 0, [_e.sent(),
                        transforms])];
                case 2:
                    if (!(operation === 'subscription')) return [3 /*break*/, 4];
                    subscriber = createSubscriber(targetSchema, rootValue, subSchemaConfig);
                    return [4 /*yield*/, subscriber({
                            document: processedRequest.document,
                            context: context,
                            variables: processedRequest.variables,
                        })];
                case 3:
                    originalAsyncIterator = (_e.sent());
                    // "subscribe" to the subscription result and map the result through the transforms
                    return [2 /*return*/, mapAsyncIterator_1.default(originalAsyncIterator, function (result) {
                            var _a;
                            var transformedResult = transforms_1.applyResultTransforms(result, transforms);
                            // wrap with fieldName to return for an additional round of resolutioon
                            // with payload as rootValue
                            return _a = {},
                                _a[info.fieldName] = transformedResult,
                                _a;
                        })];
                case 4: return [2 /*return*/];
            }
        });
    });
}
function createDocument(targetField, targetOperation, originalSelections, fragments, variables, operationName) {
    var selections = [];
    var args = [];
    originalSelections.forEach(function (field) {
        var fieldSelections = field.selectionSet
            ? field.selectionSet.selections
            : [];
        selections = selections.concat(fieldSelections);
        args = args.concat(field.arguments || []);
    });
    var selectionSet = null;
    if (selections.length > 0) {
        selectionSet = {
            kind: graphql_1.Kind.SELECTION_SET,
            selections: selections,
        };
    }
    var rootField = {
        kind: graphql_1.Kind.FIELD,
        alias: null,
        arguments: args,
        selectionSet: selectionSet,
        name: {
            kind: graphql_1.Kind.NAME,
            value: targetField,
        },
    };
    var rootSelectionSet = {
        kind: graphql_1.Kind.SELECTION_SET,
        selections: [rootField],
    };
    var operationDefinition = {
        kind: graphql_1.Kind.OPERATION_DEFINITION,
        operation: targetOperation,
        variableDefinitions: variables,
        selectionSet: rootSelectionSet,
        name: operationName,
    };
    return {
        kind: graphql_1.Kind.DOCUMENT,
        definitions: __spreadArrays([operationDefinition], fragments),
    };
}
function createExecutor(schema, rootValue, subSchemaConfig) {
    var fetcher;
    if (subSchemaConfig) {
        if (subSchemaConfig.dispatcher) {
            var dynamicLinkOrFetcher = subSchemaConfig.dispatcher(context);
            fetcher = (typeof dynamicLinkOrFetcher === 'function') ?
                dynamicLinkOrFetcher :
                linkToFetcher_1.default(dynamicLinkOrFetcher);
        }
        else if (subSchemaConfig.link) {
            fetcher = linkToFetcher_1.default(subSchemaConfig.link);
        }
        else if (subSchemaConfig.fetcher) {
            fetcher = subSchemaConfig.fetcher;
        }
        if (!fetcher && !rootValue && subSchemaConfig.rootValue) {
            rootValue = subSchemaConfig.rootValue;
        }
    }
    if (fetcher) {
        return function (_a) {
            var document = _a.document, context = _a.context, variables = _a.variables;
            return fetcher({
                query: document,
                variables: variables,
                context: { graphqlContext: context }
            });
        };
    }
    else {
        return function (_a) {
            var document = _a.document, context = _a.context, variables = _a.variables;
            return graphql_1.execute({
                schema: schema,
                document: document,
                rootValue: rootValue,
                contextValue: context,
                variableValues: variables,
            });
        };
    }
}
function createSubscriber(schema, rootValue, subSchemaConfig) {
    var link;
    if (subSchemaConfig) {
        if (subSchemaConfig.dispatcher) {
            link = subSchemaConfig.dispatcher(context);
        }
        else if (subSchemaConfig.link) {
            link = subSchemaConfig.link;
        }
        if (!link && !rootValue && subSchemaConfig.rootValue) {
            rootValue = subSchemaConfig.rootValue;
        }
    }
    if (link) {
        return function (_a) {
            var document = _a.document, context = _a.context, variables = _a.variables;
            var operation = {
                query: document,
                variables: variables,
                context: { graphqlContext: context }
            };
            var observable = apollo_link_1.execute(link, operation);
            return observableToAsyncIterable_1.observableToAsyncIterable(observable);
        };
    }
    else {
        return function (_a) {
            var document = _a.document, context = _a.context, variables = _a.variables;
            return graphql_1.subscribe({
                schema: schema,
                document: document,
                rootValue: rootValue,
                contextValue: context,
                variableValues: variables,
            });
        };
    }
}
