// graphql-tools-fork/dist/stitching/index.js
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var makeRemoteExecutableSchema_1 = __importStar(require("./makeRemoteExecutableSchema"));
exports.makeRemoteExecutableSchema = makeRemoteExecutableSchema_1.default;
exports.defaultCreateRemoteResolver = makeRemoteExecutableSchema_1.createResolver;
var introspectSchema_1 = __importDefault(require("./introspectSchema"));
exports.introspectSchema = introspectSchema_1.default;
var mergeSchemas_1 = __importDefault(require("./mergeSchemas"));
exports.mergeSchemas = mergeSchemas_1.default;
var delegateToSchema_1 = __importDefault(require("./delegateToSchema"));
exports.delegateToSchema = delegateToSchema_1.default;
var defaultMergedResolver_1 = __importDefault(require("./defaultMergedResolver"));
exports.defaultMergedResolver = defaultMergedResolver_1.default;
var createMergedResolver_1 = require("./createMergedResolver");
exports.createMergedResolver = createMergedResolver_1.createMergedResolver;
var proxiedResult_1 = require("./proxiedResult");
exports.dehoistResult = proxiedResult_1.dehoistResult;
exports.unwrapResult = proxiedResult_1.unwrapResult;
