// graphql-tools-fork/dist/stitching/resolvers.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var delegateToSchema_1 = __importDefault(require("./delegateToSchema"));
var makeMergedType_1 = require("./makeMergedType");
function generateProxyingResolvers(subschemaConfig, createProxyingResolver) {
    if (createProxyingResolver === void 0) { createProxyingResolver = defaultCreateProxyingResolver; }
    var targetSchema = subschemaConfig.schema;
    var mapping = generateSimpleMapping(targetSchema);
    var result = {};
    Object.keys(mapping).forEach(function (name) {
        result[name] = {};
        var innerMapping = mapping[name];
        Object.keys(innerMapping).forEach(function (from) {
            var _a;
            var to = innerMapping[from];
            var resolverType = to.operation === 'subscription' ? 'subscribe' : 'resolve';
            result[name][from] = (_a = {},
                _a[resolverType] = createProxyingResolver(subschemaConfig, to.operation, to.name),
                _a);
        });
    });
    return result;
}
exports.generateProxyingResolvers = generateProxyingResolvers;
function generateSimpleMapping(targetSchema) {
    var query = targetSchema.getQueryType();
    var mutation = targetSchema.getMutationType();
    var subscription = targetSchema.getSubscriptionType();
    var result = {};
    if (query) {
        result[query.name] = generateMappingFromObjectType(query, 'query');
    }
    if (mutation) {
        result[mutation.name] = generateMappingFromObjectType(mutation, 'mutation');
    }
    if (subscription) {
        result[subscription.name] = generateMappingFromObjectType(subscription, 'subscription');
    }
    return result;
}
exports.generateSimpleMapping = generateSimpleMapping;
function generateMappingFromObjectType(type, operation) {
    var result = {};
    var fields = type.getFields();
    Object.keys(fields).forEach(function (fieldName) {
        result[fieldName] = {
            name: fieldName,
            operation: operation,
        };
    });
    return result;
}
exports.generateMappingFromObjectType = generateMappingFromObjectType;
function defaultCreateProxyingResolver(subschemaConfig, operation, fieldName) {
    return function (parent, args, context, info) { return delegateToSchema_1.default({
        schema: subschemaConfig,
        operation: operation,
        fieldName: fieldName,
        args: args,
        context: context,
        info: info,
    }); };
}
function stripResolvers(schema) {
    var typeMap = schema.getTypeMap();
    Object.keys(typeMap).forEach(function (typeName) {
        if (!typeName.startsWith('__')) {
            makeMergedType_1.makeMergedType(typeMap[typeName]);
        }
    });
}
exports.stripResolvers = stripResolvers;
