// graphql-tools-fork/dist/transforms/AddMergedTypeFragments.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var AddMergedTypeFragments = /** @class */ (function () {
    function AddMergedTypeFragments(targetSchema, mapping) {
        this.targetSchema = targetSchema;
        this.mapping = mapping;
    }
    AddMergedTypeFragments.prototype.transformRequest = function (originalRequest) {
        var document = addMergedTypeFragments(this.targetSchema, originalRequest.document, this.mapping);
        return __assign(__assign({}, originalRequest), { document: document });
    };
    return AddMergedTypeFragments;
}());
exports.default = AddMergedTypeFragments;
function addMergedTypeFragments(targetSchema, document, mapping) {
    var _a;
    var typeInfo = new graphql_1.TypeInfo(targetSchema);
    return graphql_1.visit(document, graphql_1.visitWithTypeInfo(typeInfo, {
        leave: (_a = {},
            _a[graphql_1.Kind.SELECTION_SET] = function (node) {
                var parentType = typeInfo.getParentType();
                if (parentType) {
                    var parentTypeName = parentType.name;
                    var selections = node.selections;
                    if (mapping[parentTypeName]) {
                        selections = selections.concat(mapping[parentTypeName].fragment);
                    }
                    if (selections !== node.selections) {
                        return __assign(__assign({}, node), { selections: selections });
                    }
                }
            },
            _a),
    }));
}
