// graphql-tools-fork/dist/transforms/MapFields.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var MapFields = /** @class */ (function () {
    function MapFields(fieldNodeTransformerMap) {
        this.fieldNodeTransformerMap = fieldNodeTransformerMap;
    }
    MapFields.prototype.transformSchema = function (schema) {
        this.schema = schema;
        return schema;
    };
    MapFields.prototype.transformRequest = function (originalRequest) {
        var fragments = {};
        originalRequest.document.definitions.filter(function (def) { return def.kind === graphql_1.Kind.FRAGMENT_DEFINITION; }).forEach(function (def) {
            fragments[def.name.value] = def;
        });
        var document = transformDocument(originalRequest.document, this.schema, this.fieldNodeTransformerMap, fragments);
        return __assign(__assign({}, originalRequest), { document: document });
    };
    return MapFields;
}());
exports.default = MapFields;
function transformDocument(document, schema, fieldNodeTransformerMap, fragments) {
    var _a;
    if (fragments === void 0) { fragments = {}; }
    var typeInfo = new graphql_1.TypeInfo(schema);
    var newDocument = graphql_1.visit(document, graphql_1.visitWithTypeInfo(typeInfo, (_a = {},
        _a[graphql_1.Kind.SELECTION_SET] = function (node) {
            var parentType = typeInfo.getParentType();
            if (parentType) {
                var parentTypeName_1 = parentType.name;
                var newSelections_1 = [];
                node.selections.forEach(function (selection) {
                    if (selection.kind === graphql_1.Kind.FIELD) {
                        var fieldName = selection.name.value;
                        var fieldNodeTransformer = fieldNodeTransformerMap[parentTypeName_1] && fieldNodeTransformerMap[parentTypeName_1][fieldName];
                        var transformedSelection = fieldNodeTransformer
                            ? fieldNodeTransformer(selection, fragments)
                            : selection;
                        if (Array.isArray(transformedSelection)) {
                            newSelections_1 = newSelections_1.concat(transformedSelection);
                        }
                        else if (transformedSelection.kind === graphql_1.Kind.FIELD) {
                            newSelections_1.push(transformedSelection);
                        }
                        else {
                            newSelections_1.push(transformedSelection);
                        }
                    }
                    else {
                        newSelections_1.push(selection);
                    }
                });
                return __assign(__assign({}, node), { selections: newSelections_1 });
            }
        },
        _a)));
    return newDocument;
}
