// graphql-tools-fork/dist/transforms/RenameTypes.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var isSpecifiedScalarType_1 = __importDefault(require("../utils/isSpecifiedScalarType"));
var Interfaces_1 = require("../Interfaces");
var utils_1 = require("../utils");
var RenameTypes = /** @class */ (function () {
    function RenameTypes(renamer, options) {
        this.renamer = renamer;
        this.map = {};
        this.reverseMap = {};
        var _a = options || {}, _b = _a.renameBuiltins, renameBuiltins = _b === void 0 ? false : _b, _c = _a.renameScalars, renameScalars = _c === void 0 ? true : _c;
        this.renameBuiltins = renameBuiltins;
        this.renameScalars = renameScalars;
    }
    RenameTypes.prototype.transformSchema = function (originalSchema) {
        var _a;
        var _this = this;
        return utils_1.visitSchema(originalSchema, (_a = {},
            _a[Interfaces_1.VisitSchemaKind.TYPE] = function (type) {
                if (isSpecifiedScalarType_1.default(type) && !_this.renameBuiltins) {
                    return undefined;
                }
                if (type instanceof graphql_1.GraphQLScalarType && !_this.renameScalars) {
                    return undefined;
                }
                var oldName = type.name;
                var newName = _this.renamer(oldName);
                if (newName && newName !== oldName) {
                    _this.map[oldName] = type.name;
                    _this.reverseMap[newName] = oldName;
                    var newType = utils_1.cloneType(type);
                    newType.name = newName;
                    return newType;
                }
            },
            _a[Interfaces_1.VisitSchemaKind.ROOT_OBJECT] = function (type) {
                return undefined;
            },
            _a));
    };
    RenameTypes.prototype.transformRequest = function (originalRequest) {
        var _a;
        var _this = this;
        var newDocument = graphql_1.visit(originalRequest.document, (_a = {},
            _a[graphql_1.Kind.NAMED_TYPE] = function (node) {
                var name = node.name.value;
                if (name in _this.reverseMap) {
                    return __assign(__assign({}, node), { name: {
                            kind: graphql_1.Kind.NAME,
                            value: _this.reverseMap[name],
                        } });
                }
            },
            _a));
        return {
            document: newDocument,
            variables: originalRequest.variables,
        };
    };
    RenameTypes.prototype.transformResult = function (result) {
        return __assign(__assign({}, result), { data: this.renameTypes(result.data) });
    };
    RenameTypes.prototype.renameTypes = function (value) {
        var _this = this;
        if (value == null) {
            return value;
        }
        else if (Array.isArray(value)) {
            value.forEach(function (v, index) {
                value[index] = _this.renameTypes(v);
            });
            return value;
        }
        else if (typeof value === 'object') {
            Object.keys(value).forEach(function (key) {
                value[key] = key === '__typename' ?
                    _this.renamer(value[key]) : _this.renameTypes(value[key]);
            });
            return value;
        }
        else {
            return value;
        }
    };
    return RenameTypes;
}());
exports.default = RenameTypes;
